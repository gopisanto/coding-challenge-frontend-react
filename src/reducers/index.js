import { combineReducers } from 'redux';
import axios from 'axios';
import { get, isEmpty } from 'lodash';
import Immutable from 'seamless-immutable';

import getFilteredBikes from '../helpers';

export const INITIAL_STATE = Immutable({
  bikes: [],
  loading: false,
  filteredBikes: [],
  search: '',
  startDate: '',
  endDate: '',
  error: null,
});

export const UPDATE_BIKES = 'reducer.updateBikes';
export const UPDATE_LOADER = 'reducer.updateLoader';
export const UPDATE_BIKE_INCIDENT = 'reducer.updateBikeIncident';
export const FILTER_BIKES_LIST = 'reducer.filterBikeList';
export const SERVICE_ERROR = 'reducer.updateServiceError';

export const updateBikeIncident = payload => ({
  type: UPDATE_BIKE_INCIDENT,
  payload,
});

export const updateBikes = payload => ({
  type: UPDATE_BIKES,
  payload,
});

export const updateLoader = payload => ({
  type: UPDATE_LOADER,
  payload,
});

export const filterBikeList = payload => ({
  type: FILTER_BIKES_LIST,
  payload,
});

export const updateServiceError = payload => ({
  type: SERVICE_ERROR,
  payload,
});

export const fetchBikes = () => async (dispatch, getState) => {
  if (!isEmpty(get(getState(), 'bikes.bikes'))) {
    return;
  }
  dispatch(updateServiceError(null));
  dispatch(updateLoader(true));
  try {
    const bikes = await axios.get('https://bikewise.org:443/api/v2/incidents');

    dispatch(updateBikes(bikes.data.incidents));
  } catch (err) {
    dispatch(updateServiceError(`${get(err, 'response.data.error')}. Try refreshing.` || 'Some issue occured please try refreshing.'));
  }

  dispatch(updateLoader(false));
};

export const fetchBikeDetail = incidentId => async (dispatch, getState) => {
  let bike = get(getState(), `bikes.incidents.${incidentId}`);

  if (!isEmpty(bike)) {
    return;
  }

  dispatch(updateServiceError(null));
  dispatch(updateLoader(true));
  try {
    bike = await axios.get(`https://bikewise.org:443/api/v2/incidents/${incidentId}`);
    dispatch(updateBikeIncident({ incident: bike.data.incident, incidentId }));
  } catch (err) {
    dispatch(updateServiceError(get(err, 'response.data.error')));
  }
  dispatch(updateLoader(false));
};

export const bikes = (state = INITIAL_STATE, action) => {
  const { type, payload } = action;

  switch (type) {
    case UPDATE_BIKES: return state
      .set('bikes', payload)
      .set('filteredBikes', payload)
      .set('error', null);
    case UPDATE_LOADER: return state.set('loading', payload);
    case UPDATE_BIKE_INCIDENT: {
      const { incident, incidentId } = payload;
      return state
        .update('incidents', incidents => Object.assign({}, incidents, { [incidentId]: incident }))
        .set('error', null);
    }
    case FILTER_BIKES_LIST: return state.set(
      'filteredBikes',
      getFilteredBikes(Object.assign({}, { bikes: state.bikes }, payload)),
    )
      .set('search', payload.search)
      .set('startDate', payload.startDate)
      .set('endDate', payload.endDate);
    case SERVICE_ERROR: return state.set('error', payload);
    default: return state;
  }
};

export default combineReducers({
  bikes,
});
