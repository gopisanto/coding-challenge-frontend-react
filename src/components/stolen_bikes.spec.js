import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16.1';
import configureMockStore from 'redux-mock-store';
import renderer from 'react-test-renderer';
import thunk from 'redux-thunk';

import StolenBikes from './stolen_bikes';
import { INITIAL_STATE, updateLoader, fetchBikes } from '../reducers';

Enzyme.configure({ adapter: new Adapter() });

const bikes = [
  {
    id: 1,
    description: 'Hero ranger is stolen at berlin station.',
    occurred_at: 1550012400, //13th Feb 2019
  },
  {
    id: 2,
    description: 'Blue byke is stolen at my home.',
    occurred_at: 1550185200, // 15th Feb 2019
  },
  {
    id: 3,
    description: 'Byke is stolen from my hand.',
    occurred_at: 1550271600, // 16th Feb 2019
  },
];

describe('<StolenBikes />', () => {
  let store, container;

  beforeEach(() => {
    store = configureMockStore([thunk])({ bikes: [] });
    container = renderer.create(<StolenBikes store={store} />).toJSON();
  });

  test('renders loading screen', () => {
    expect(container).toMatchSnapshot();
    store.dispatch(updateLoader(true));
    expect(container).toMatchSnapshot();
  });

  test('renders bikes list', () => {
    expect(container).toMatchSnapshot();

    return store.dispatch(fetchBikes()).then(() => {
      expect(container).toMatchSnapshot();
    });
  })
});
