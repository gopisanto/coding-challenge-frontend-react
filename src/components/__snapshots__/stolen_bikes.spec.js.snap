// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StolenBikes /> renders bikes list 1`] = `
<div
  className="stolen-bikes"
>
  <div
    className="header"
  >
    <img
      alt="police"
      className="logo"
      src="police.jpg"
    />
    <div
      className="titles"
    >
      <h1>
        Police Department Of Berlin
      </h1>
      <h3>
        Stolen Bykes
      </h3>
    </div>
  </div>
  <div
    className="search-bar"
  >
    <input
      onChange={[Function]}
      placeholder="search case descriptions"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="from"
      type="date"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="to"
      type="date"
      value=""
    />
    <input
      disabled={true}
      onClick={[Function]}
      type="button"
      value="Find cases"
    />
  </div>
  <div>
    No data found
  </div>
</div>
`;

exports[`<StolenBikes /> renders bikes list 2`] = `
<div
  className="stolen-bikes"
>
  <div
    className="header"
  >
    <img
      alt="police"
      className="logo"
      src="police.jpg"
    />
    <div
      className="titles"
    >
      <h1>
        Police Department Of Berlin
      </h1>
      <h3>
        Stolen Bykes
      </h3>
    </div>
  </div>
  <div
    className="search-bar"
  >
    <input
      onChange={[Function]}
      placeholder="search case descriptions"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="from"
      type="date"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="to"
      type="date"
      value=""
    />
    <input
      disabled={true}
      onClick={[Function]}
      type="button"
      value="Find cases"
    />
  </div>
  <div>
    No data found
  </div>
</div>
`;

exports[`<StolenBikes /> renders loading screen 1`] = `
<div
  className="stolen-bikes"
>
  <div
    className="header"
  >
    <img
      alt="police"
      className="logo"
      src="police.jpg"
    />
    <div
      className="titles"
    >
      <h1>
        Police Department Of Berlin
      </h1>
      <h3>
        Stolen Bykes
      </h3>
    </div>
  </div>
  <div
    className="search-bar"
  >
    <input
      onChange={[Function]}
      placeholder="search case descriptions"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="from"
      type="date"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="to"
      type="date"
      value=""
    />
    <input
      disabled={true}
      onClick={[Function]}
      type="button"
      value="Find cases"
    />
  </div>
  <div>
    No data found
  </div>
</div>
`;

exports[`<StolenBikes /> renders loading screen 2`] = `
<div
  className="stolen-bikes"
>
  <div
    className="header"
  >
    <img
      alt="police"
      className="logo"
      src="police.jpg"
    />
    <div
      className="titles"
    >
      <h1>
        Police Department Of Berlin
      </h1>
      <h3>
        Stolen Bykes
      </h3>
    </div>
  </div>
  <div
    className="search-bar"
  >
    <input
      onChange={[Function]}
      placeholder="search case descriptions"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="from"
      type="date"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="to"
      type="date"
      value=""
    />
    <input
      disabled={true}
      onClick={[Function]}
      type="button"
      value="Find cases"
    />
  </div>
  <div>
    No data found
  </div>
</div>
`;
